### PR TITLE
Don't rehash blocks in flush. Do it in extent open

### PR DESCRIPTION
**Pull request by faithanalog**
A few weeks ago, the team noticed a few edge cases in crash consistency that
were addressed by associating a hash of the on-disk data with each block
context. With that change, extent flush function rehashed the entire
extent during a flush, and ran a sqlite DELETE query for every single
block in the extent. This was rather slow.

This commit introduces an in-memory cache of `on_disk_hash`es for
the latest write to each block since the last flush. That is to say, if
a write is issued for blocks `0..=5` and blocks `2..=7`, we store 8
hashes, one for each of blocks `0..=7`. Later hashes shadow previous
hashes in this in-memory cache. This cache is cleared during a flush.

All of the hashes are still present in the sqlite database prior to the
flush; that behavior is preserved.

When it comes time to flush, we use our in-memory cache of hashes to
execute `DELETE` statements for only the blocks that have changed since
the previous flush. Additionally, since we kept the hash around from the
last write, we don't need to re-read that data from disk. This operates
under the assumption that because our process has not crashed since the
last flush, the data on disk after the fsync is the same as the data we
wrote.

This optimization reduces the cost of a flush a minor amount if the
entire extent is dirty, and reduces the cost significantly if only a few
blocks are dirty. It is *roughly* linear with the number of dirty
blocks. Flushes that took many milliseconds now do not.

When recovering from a crash, this optimization means that flushes will
not automatically clear out invalid contexts for blocks written between a
flush() and the crash. Therefore this commit introduces a new function:
`Inner::fully_rehash_and_clean_all_stale_contexts`. As the name implies,
this function performs the behavior that the old flush did, rehashing
the entire contents of the extent on disk and clearing out any invalid
contexts from the database. `Extent::open` calls this function.

This rehashing only happens if the extent is marked dirty, as the extent
is marked dirty before any contexts are written. This optimization keeps
startup performance impact to a minimum. This rehashing does not then
mark the extent as not-dirty when it's done. Maybe it should mark the
extent as not-dirty. Please comment on this!

### CODE REVIEW

**Comment by leftwo** [comment]
I crafted up two dtrace scripts (attached below) to measure what goes on during a flush to see
the impact here.  I created an ubuntu VM in Omicron and attached a 2nd disk to it where I ran some 
FIO workloads.  Then I measured the flush times using dtrace on a specific downstairs.

I'm measuring time in nanoseconds here for time we take in each phase of the flush operation.

`flush submit-extent`: how long the downstairs takes once it pulls a job off the queue
to get to the point where it is farming that flush out to each extent.
`extent data flush`: the total time all extents took to fsflush the data file.  I gather this
but adding up the time each extent took for a specific flush, then entering the total time across
all extents (for a specific flush) and recording that as a single point.
`extent rehash`: is (total time like above) how long we spent determining the data checksum
we use for each block in the extent (this is the code path that changes)
`extent sqlite`: is (total time like above) how long we spent updating sqlite with our new
checksums.
`extent seek`: is the final flush step, updating flush# and gen# for an extent and committing
that to the database.
`flush total all extents`: the time taken for all the `extent *` above, recorded for a single
flush.  Basically the time delta from when we started looping over all extents till the time we
had completed all extents.
`extent-done`: the time from completing the flush on all extents to the time we
are just going to send the result back to the upstairs.


First, the numbers using stock (before this PR) bits:

```
    21021  flush submit-extent
           value  ------------- Distribution ------------- count
            4096 |                                         0
            8192 |@@@@                                     11
           16384 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@        94
           32768 |@@                                       6
           65536 |                                         0
          131072 |                                         0
          262144 |                                         1
          524288 |                                         1
         1048576 |                                         0
         2097152 |                                         0
         4194304 |                                         1
         8388608 |                                         0

    21021  extent data flush
           value  ------------- Distribution ------------- count
            4096 |                                         0
            8192 |@@                                       5
           16384 |@@@@                                     10
           32768 |@@                                       6
           65536 |@@@@                                     12
          131072 |@@@@@@@@@@@@                             33
          262144 |@                                        3
          524288 |                                         0

    21021  extent rehash
           value  ------------- Distribution ------------- count
         8388608 |                                         0
        16777216 |@@@@@@@@@@                               29
        33554432 |@@@@@@                                   17
        67108864 |@@@@@@                                   16
       134217728 |@                                        4
       268435456 |@                                        3
       536870912 |                                         0

    21021  extent sqlite
           value  ------------- Distribution ------------- count
        33554432 |                                         0
        67108864 |@                                        2
       134217728 |@@@@@@@@@@@                              30
       268435456 |@@@@@@@@                                 23
       536870912 |@@@@@                                    14
      1073741824 |                                         0

    21021  extent seek
           value  ------------- Distribution ------------- count
           65536 |                                         0
          131072 |                                         1
          262144 |@@@@@@@@@@                               29
          524288 |@@@@@@@                                  21
         1048576 |@@@@@@                                   16
         2097152 |@                                        2
         4194304 |                                         0

    21021  flush total all extents
           value  ------------- Distribution ------------- count
         1048576 |                                         0
         2097152 |                                         1
         4194304 |@@@@@@@@@@@@@@@                          44
         8388608 |                                         0
        16777216 |                                         0
        33554432 |                                         0
        67108864 |                                         0
       134217728 |@@@@@@@@@@@                              30
       268435456 |@@@@@@@@                                 24
       536870912 |@@@@@                                    13
      1073741824 |@                                        2
      2147483648 |                                         0

    21021  flush extent-done
           value  ------------- Distribution ------------- count
            4096 |                                         0
            8192 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@             81
           16384 |@@@@@@@@@@@@                             33
           32768 |                                         0
```

You can see there are some flush commands that take up to a second!, with a many up to 1/2 a second.


Now, here are dtrace numbers using a new VM (omicron, ubuntu, fio, etc) running the new bits:
```
     1448  flush submit-extent
           value  ------------- Distribution ------------- count
            4096 |                                         0
            8192 |@@@                                      9
           16384 |@@@@@@@@@@@@@@@@@@@@                     63
           32768 |@@@@@@@@@@@@@@@@@@                       56

     1448  extent data flush
           value  ------------- Distribution ------------- count
            4096 |                                         0
            8192 |@                                        2
           16384 |@@                                       5
           32768 |                                         1
           65536 |@@@@                                     14
          131072 |@@@@@@@@@@@@@@@@@@@@@@                   69
          262144 |@@@@@                                    17
          524288 |                                         0

     1448  extent rehash
           value  ------------- Distribution ------------- count
            4096 |                                         0
            8192 |@@@@@@@@@@                               31
           16384 |@@@                                      9
           32768 |@@@                                      10
           65536 |@                                        3
          131072 |@                                        3
          262144 |@                                        2
          524288 |@@                                       6
         1048576 |@@@@@@@@@@@@@                            43
         2097152 |                                         1
         4194304 |                                         0

     1448  extent sqlite
           value  ------------- Distribution ------------- count
           65536 |                                         1
          131072 |@@@@                                     12
          262144 |@@@@@@@                                  23
          524288 |@@                                       7
         1048576 |@@                                       5
         2097152 |@@                                       5
         4194304 |@                                        4
         8388608 |                                         1
        16777216 |@@@@@@@@@@@@@@@                          49
        33554432 |                                         1
        67108864 |                                         0

     1448  extent seek
           value  ------------- Distribution ------------- count
           65536 |                                         0
          131072 |@                                        4
          262144 |@@@@@@@@@@@@@@@@@@@@@                    68
          524288 |@@@@@@@@@@                               32
         1048576 |@                                        4
         2097152 |                                         0


     1448  flush total all extents
           value  ------------- Distribution ------------- count
         2097152 |                                         0
         4194304 |@@@@@@@@@@@@@@@@@@@@@                    67
         8388608 |@@@                                      11
        16777216 |@@@                                      9
        33554432 |@@@@@@@@@@@@@                            41
        67108864 |                                         0


     1448  flush extent-done
           value  ------------- Distribution ------------- count
            4096 |                                         0
            8192 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@      111
           16384 |@@@@@                                    17
           32768 |                                         0
```

So, we can see the worst case total time for all extents is much much lower, as well as several of the
phases in the flush are now shorter.

**Comment by leftwo** [comment]
DTrace script for the current flush path (before this PR)
[perf-downstairs-flush-tick.d.txt](https://github.com/oxidecomputer/crucible/files/10282569/perf-downstairs-flush-tick.d.txt)

The same DTrace script, but updated to work after this PR.
[perf-downstairs-flush-tick-new.d.txt](https://github.com/oxidecomputer/crucible/files/10282571/perf-downstairs-flush-tick-new.d.txt)

File: downstairs/src/region.rs [hunk 1]
```diff
@@ -1270,6 +1346,75 @@ impl Extent {
...
+        // side-effects. There are no unhandled errors here.
+        drop(inner_file_buffered);
+
+        inner.truncate_encryption_contexts_and_hashes(
+            extent_block_indexes_and_hashes,
+        )?;
+
+        inner.file.seek(SeekFrom::Start(0))?;
+
+        // XXX should we be clearing the dirty flag here? If we do, should we
```
Comment (leftwo) [inline downstairs/src/region.rs:1410]: No clearing of dirty bit here, we want the repair to know that this block was dirty.

**Review by leftwo — COMMENTED** [review]
It's clear the flush performance is improved, but we need to look at the write-unwritten path and make sure that still works.

File: downstairs/src/region.rs [hunk 2]
```diff
@@ -1270,6 +1346,75 @@ impl Extent {
...
+    /// that do not correlate to data currently stored on disk. This is
+    /// primarily when opening an extent after recovering from a crash, since
+    /// irrelevant hashes will normally be cleared out during a flush().
+    ///
+    /// By default this function will only do work if the extent is marked
+    /// dirty. Set `force_override_dirty` to `true` to override this behavior.
+    /// This override should generally not be necessary, as the dirty flag
+    /// is set before any contexts are written.
+    #[instrument]
+    fn fully_rehash_and_clean_all_stale_contexts(
```
[the remark this answers is not in the document]
Comment (faithanalog) [reply downstairs/src/region.rs:1360]: No. Just as with the previous implementation of flush(), this only clears out contexts that do not match any existing on-disk block.

File: downstairs/src/region.rs [hunk 3]
```diff
@@ -604,13 +612,23 @@ impl Extent {
...
+            inner: Some(Mutex::new(Inner {
+                file,
+                metadb,
+                dirty_blocks: HashMap::new(),
+            })),
+        };
+
+        // Clean out any irrelevant block contexts, which may be present if downstairs
+        // crashed between a write() and a flush().
+        extent.fully_rehash_and_clean_all_stale_contexts(false)?;
```
[the remark this answers is not in the document]
Comment (faithanalog) [reply downstairs/src/region.rs:629]: This will remove any contexts not associated with any data on disk. So yeah this could result in a change in the data. To cause this to happen you would need to

- write data to a read-write region
- crash or otherwise shut down downstairs before a flush succeeds
- mark it as read-only before opening the region again
- open the region

File: downstairs/src/region.rs [hunk 4]
```diff
@@ -251,23 +252,29 @@ impl Inner {
...
         extent_block_indexes_and_hashes: Vec<(usize, u64)>,
     ) -> Result<()> {
         let tx = self.metadb.transaction()?;
 
         let stmt = "DELETE FROM block_context where block == ?1 and on_disk_hash != ?2";
+        let mut stmt = tx.prepare_cached(stmt)?;
+
+        let n_blocks = extent_block_indexes_and_hashes.len();
 
+        cdt::extent__context__truncate__start!(|| n_blocks as u64);
```
[the remark this answers is not in the document]
Comment (faithanalog) [reply downstairs/src/region.rs:270]: it might make sense to include the commit. I was mainly just using this as debug when figuring out that sorting made a big difference and I left it in.

File: downstairs/src/region.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (leftwo) [reply downstairs/src/region.rs:1360]: Do we have a test for that case?  Maybe re-use the `multiple_context` test in `downstairs/src/region.rs` that
* Creates a region.
* Sets a few inner.contexts on some blocks, but nothing for others
* Create a full vec `extent_block_indexes_and_hashes` with hash data in every block
* Call `truncate_encryption_contexts_and_hashes` with the created vec above.
* Verify afterwards that the only contexts that exist are for the few we originally sent, that no other contexts exist.

I originally thought we would have to do this all in the integration suite area, but reading more I think we can
add a test around `truncate_encryption_contexts_and_hashes` that will verify this for us, and prevent
future us (most likely me) from doing something to break this required behavior.

[the remark this answers is not in the document]
Comment (faithanalog) [reply downstairs/src/region.rs:1360]: yeah so i actually modified `test_flush_removes_partial_writes` to test this instead. it can stand for a rename though. This makes sense to me because we're intentionally removing the behavior where flush removes partial writes, and moving it into the fully_rehash thing.

right now the changes i made have it close all the extents and re-open them to make sure it happens on extent open but we could flip that to an explicit call to `fully_rehash_...`

[the remark this answers is not in the document]
Comment (leftwo) [reply downstairs/src/region.rs:1360]: Ah, yeah, `test_flush_removes_partial_writes()` is pretty good, but I think we can make it just
a little better, either in the same test or on another test. In addition to what there is currently, have
a test verify that: 
A totally unwritten block is still "unwritten", meaning a call to `region_write` with `only_write_unwritten` = `true`
will indeed write to the block.
A written block with multiple contexts is truncated to a single context, then we verify that a write_unwritten will not write to that block.  This would be the "normal" case where the truncate does what we want it to do.

Sorry for adding all this.   I'm very nervous about breaking something without knowing and having all these tests
help me sleep at night.

File: downstairs/src/region.rs [hunk 5]
```diff
@@ -1093,23 +1121,41 @@ impl Extent {
...
+            // into that function too. This might be nice, since then a block
+            // context could never be set without marking the block as dirty.
+            // On the other paw, our test suite very much likes the fact that
+            // tx_set_block_context would mark blocks as dirty, since it lets
+            // us easily test specific edge-cases of the database state.
+            // Could make another function that wraps tx_set_block_context
+            // and handles this as well.
+            let _ = inner
+                .dirty_blocks
+                .insert(write.offset.value as usize, on_disk_hash);
```
Comment (jmpesp) [inline downstairs/src/region.rs:1155]: I think there's a consistency problem here: if any of the `write_all` calls below return Err, the value in `dirty_blocks` at a particular offset will be wrong. If the Upstairs then sends a Flush, we'll be removing the wrong context rows.

**Review by jmpesp — CHANGES_REQUESTED** [review]

File: downstairs/src/region.rs (hunk 5 above)
Comment (faithanalog) [reply downstairs/src/region.rs:1153]: I think you're right about that, let me adjust. I wonder if there's some way I can force an error to test for this situation.

**Comment by faithanalog** [comment]
i reworked it to re-read data from disk to rehash, but to only rehash blocks that may have changed since last flush (anything where a write command was received). consequently we lose the speed benefits that my previous version had during full-extent flushes, but we don't go any slower for full flushes than in `main`, and we retain the majority of the benefits for any flush smaller than a full extent.

File: downstairs/src/region.rs [hunk 6]
```diff
@@ -1600,25 +1600,37 @@ impl Region {
...
-            })
-            .collect::<Result<Vec<Extent>>>()?;
+            });
+            these_extent_handles.push(handle);
+        }
+
+        let mut these_extents = Vec::with_capacity(these_extent_handles.len());
+        for handle in these_extent_handles  {
+            these_extents.push(handle.await??);
+        }
```
Comment (faithanalog) [inline downstairs/src/region.rs:1628]: here's the bit where im tokio::spawning

**Comment by faithanalog** [comment]
alright im watching the jobs here and i think the parallel thing is like stochastically not working so Im gonna back that one out. it was worth a shot.

**Comment by faithanalog** [comment]
> alright im watching the jobs here and i think the parallel thing is like stochastically not working so Im gonna back that one out. it was worth a shot.

oh you know what I bet you 50 diamonds in minecraft that the reason it was failing is because creating the seed sqlite database happens inside of extent creation, rather than outside of it. so multiple jobs could

- see seed.db doesn't exist
- try to create it
- stomp on each other

or

- see seed.db exists, but seed.db isn't done being created
- copy in the incorrect db

I'll save that for another PR, but if we want to parallelize extent creation/open, that's the place to look I think

**Review by jmpesp — CHANGES_REQUESTED** [review]
I tested this on my cluster of local machines, and saw a **stark** difference in some real-world workloads (`apt update`, `apt install nginx`, etc). Turns out not re-reading the whole region is a good idea :) Besides some additional tests this should go in ASAP.

File: downstairs/src/region.rs [hunk 7]
```diff
@@ -1219,54 +1289,203 @@ impl Extent {
...
+
+        // All blocks that may have changed since the last flush
+        let mut dirty_blocks: Vec<usize> = inner.dirty_blocks.drain().collect();
+
+        // Sorting here is extremely important to get good performance out of
+        // truncate_encryption_contexts_and_hashes() later. It also means we
+        // can efficiently re-read only the dirty parts of the file.
+        dirty_blocks.sort_unstable();
+
+        // Rehash any parts of the file that we wrote data to since the last
```
[the remark this answers is not in the document]
Comment (faithanalog) [reply downstairs/src/region.rs:1319]: what are you thinking for unit tests here? I'm figuring you mean like trying writes to some disjoint ranges of blocks and then flushing to make sure that works right, anything else?

[the remark this answers is not in the document]
Comment (jmpesp) [reply downstairs/src/region.rs:1319]: exactly yah - exercising edge cases like disjoint ranges, filling up the range buffer, etc.

**Comment by faithanalog** [comment]
rebased and added `test_flush_after_multiple_disjoint_writes()`

**Comment by faithanalog** [comment]
@jmpesp  let me know if you want any other tests. i know you mentioned filling the range buffer? i assume you just mean like, writing an entire extent? in which case, crudd is basically already testing that but i can put in an explicit one

File: downstairs/src/region.rs [hunk 8]
```diff
@@ -4350,11 +4804,179 @@ mod test {
...
+            // Check that they're right.
+            assert_eq!(expected_ctxts, actual_ctxts);
+        }
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    /// We need to make sure that a flush will properly adjust the DB hashes
+    /// after issuing multiple writes to different disconnected sections of
```
Comment (jmpesp) [inline downstairs/src/region.rs:4899]: this comment needs updating :)

**Review by jmpesp — APPROVED** [review]
Let's :rocket: this, it's a huge improvement

**Review by leftwo — APPROVED** [review]
This change enabled completion of a perf test I was trying.
The stock bits could not even finish the test, as flushes were taking longer than 50 seconds and the upstairs would timeout and disconnect. 
With these changes, the test can actually finish!
